pyem/metadata.py

```diff
@@ -292,23 +292,80 @@ def cryosparc_2_cs_ctf_parameters(cs, df=None):
     log = logging.getLogger('root')
     if df is None:
         df = pd.DataFrame()
-    if 'ctf/tilt_A' in cs.dtype.names:
-        log.debug("Recovering beam tilt and converting to mrad")
-        df[star.Relion.BEAMTILTX] = np.arcsin(cs['ctf/tilt_A'][:, 0] / cs['ctf/cs_mm'] * 1e-7) * 1e3
-        df[star.Relion.BEAMTILTY] = np.arcsin(cs['ctf/tilt_A'][:, 1] / cs['ctf/cs_mm'] * 1e-7) * 1e3
-    if 'ctf/shift_A' in cs.dtype.names:
-        pass
-    if 'ctf/trefoil_A' in cs.dtype.names:
-        pass
-        # df[star.Relion.ODDZERNIKE] = cs['ctf/trefoil_A']
-    if 'ctf/tetrafoil_A' in cs.dtype.names:
-        pass
-        # df[star.Relion.EVENZERNIKE] = cs['ctf/tetra_A']
+    if 'ctf/accel_kv' in cs.dtype.names:
+        # Calculate electron wavelentgh
+        # https://www.jeol.co.jp/en/words/emterms/search_result.html?keyword=wavelength%20of%20electron
+        E = cs['ctf/accel_kv']  # convert units?
+        wavelength = 1.23e3 / np.sqrt(E*(1 + 9.78e-7 * E))  # ? TODO: check this
+    else:
+        wavelength = None
     if 'ctf/anisomag' in cs.dtype.names:
         df[star.Relion.MAGMAT00] = cs['ctf/anisomag'][:, 0]
         df[star.Relion.MAGMAT01] = cs['ctf/anisomag'][:, 1]
         df[star.Relion.MAGMAT10] = cs['ctf/anisomag'][:, 2]
         df[star.Relion.MAGMAT11] = cs['ctf/anisomag'][:, 3]
+    if 'ctf/tilt_A' in cs.dtype.names:
+        log.debug("Recovering beam tilt and converting to mrad")
+        df[star.Relion.BEAMTILTX] = np.arcsin(cs['ctf/tilt_A'][:, 0] / cs['ctf/cs_mm'] * 1e-7) * 1e3
+        df[star.Relion.BEAMTILTY] = np.arcsin(cs['ctf/tilt_A'][:, 1] / cs['ctf/cs_mm'] * 1e-7) * 1e3
+        df[star.Relion.Z_neg1_3] = df[star.Relion.BEAMTILTX]  # does this also require a conversion factor of 2π * Cs * λ^2
+        df[star.Relion.Z_pos1_3] = df[star.Relion.BEAMTILTY]  # does this also require a conversion factor of 2π * Cs * λ^2
```

Comment:
Does this also need a conversion factor, or is that already taken care of in BEAMTILTX and BEAMTILTY?

Reply:
I think the _3 coefficients are 2 * π * λ**2 times the tilt vector (`tilt_A`) like in Eqn 5 from the older "new tools for automated high-resolution..." paper.

Reply:
> I think the _3 coefficients are 2 * π * λ**2 times the tilt vector (tilt_A) like in Eqn 5 from the older "new tools for automated high-resolution..." paper.

Did you mean to write `2 * π * Cs * λ**2` here instead? Equation 5 includes the Cs (spherical abberation) term, and we also used `2 * π * Cs * λ**2` as the prefactor for the trefoil terms in the code here too.

I'm going to proceed on the assumption that you did mean to say `2 * π * Cs * λ**2`, so if that's not the case please clarify and I can edit/fix things.

Reply:
... although I see here that the `cs['ctf/cs_mm']` term was divided out from `ctf/tilt_A`, so perhaps i should not be re-introducing it.